clients/client-py/test/test_client.py

```diff
@@ -618,49 +639,51 @@ def test_builds_surl_keyword(self):
 verifies that we can call real methods with both position and keyword
 args"""
 
-def setUp(self):
-    self.fakeResponse = ''
-
-    def fakeSite(url, request):
-        self.gotUrl = urllib.parse.urlunsplit(url)
-        self.gotRequest = request
-        return self.fakeResponse
-    self.fakeSite = fakeSite
-
-def test_no_args_no_input(self):
-    with httmock.HTTMock(self.fakeSite):
-        self.client.no_args_no_input()
-    assert self.gotUrl == 'https://tc-tests.example.com/api/fake/v1/no_args_no_input'
-
-def test_two_args_no_input(self):
-    with httmock.HTTMock(self.fakeSite):
-        self.client.two_args_no_input('1', '2')
-    assert self.gotUrl == 'https://tc-tests.example.com/api/fake/v1/two_args_no_input/1/2'
-
-def test_no_args_with_input(self):
-    with httmock.HTTMock(self.fakeSite):
-        self.client.no_args_with_input({'x': 1})
-    assert self.gotUrl == 'https://tc-tests.example.com/api/fake/v1/no_args_with_input'
-    assert json.loads(self.gotRequest.body) == {"x": 1}
-
-def test_no_args_with_empty_input(self):
-    with httmock.HTTMock(self.fakeSite):
-        self.client.no_args_with_input({})
-    assert self.gotUrl == 'https://tc-tests.example.com/api/fake/v1/no_args_with_input'
-    assert json.loads(self.gotRequest.body) == {}
-
-def test_two_args_with_input(self):
-    with httmock.HTTMock(self.fakeSite):
-        self.client.two_args_with_input('a', 'b', {'x': 1})
-    assert self.gotUrl == 'https://tc-tests.example.com/api/fake/v1/two_args_with_input/a/b'
-    assert json.loads(self.gotRequest.body) == {"x": 1}
-
-def test_kwargs(self):
-    with httmock.HTTMock(self.fakeSite):
-        self.client.two_args_with_input(
+@pytest.yield_fixture(scope='function')
+def fakeResponse(fakeResponse):
+    yield fakeResponse
+
+@pytest.fixture(scope='function')
+def fakeSite(url, request, fakeResponse):
+    gotUrl = urllib.parse.urlunsplit(url)
+    gotRequest = request
+    return fakeResponse
+
+
+def test_no_args_no_input(client, fakeSite):
```

Comment:
From this test case, the next five test are are failing. I tried to make a fixture for `fakeSite & fakeResponse` but I think the test `test_no_args_no_input` can't resolve the `fakeSite`.

I'm getting this error : 

![Screenshot (251)](https://user-images.githubusercontent.com/21038543/64163181-68158800-ce5e-11e9-8969-b4986c8b465b.png)

the `fakeSite` fixture can't resolve `url`, I don't know why. 😕
Could you please help me here?

PS : I think rest of the files are fixed.

Thanks a lot! :)

Reply:
I believe I fixed this. I can show you the patch if you want, but I basically:

1. `import functools`
2. change `fakeSite` from a fixture, where all the args and kwargs need to be fixtures, to a function with kwargs we can fill in:

```
def fakeSite(url, request, expected_url=None, expected_body=None):
    if expected_url is not None:
        assert urllib.parse.urlunsplit(url) == expected_url
    if expected_body is not None:
        assert json.loads(request.body) == expected_body
    return ""
```

Then, to use:

```
def test_no_args_no_input(client):
    site = functools.partial(
        fakeSite,
        expected_url='https://tc-tests.example.com/api/fake/v1/no_args_no_input',
        # if we have an expected body, also pass `expected_body={...},`
    )
    with httmock.HTTMock(site):
         client.no_args_no_input()
```

For the `json.loads` values, set `expected_body`. Essentially, we're creating another function based on `fakeSite`, but with `expected_body` and `expected_url` optionally pre-populated.

After I greened this up, we get a bunch of linting errors from flake8... those will need cleaning up too. Luckily that tells us what needs fixing on which line, and how to fix.